src/H5Fio.c

```diff
@@ -327,6 +327,67 @@ H5F_shared_select_write(H5F_shared_t *f_sh, H5FD_mem_t type, uint32_t count, H5S
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5F_shared_select_write() */
 
+herr_t
+H5F_shared_vector_read(H5F_shared_t *f_sh, uint32_t count, H5FD_mem_t types[], haddr_t addrs[],
+                       size_t sizes[], void *bufs[])
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    /* Sanity checks */
+    HDassert(f_sh);
+    HDassert((types) || (count == 0));
+    HDassert((addrs) || (count == 0));
+    HDassert((sizes) || (count == 0));
+    HDassert((bufs) || (count == 0));
+
+    /* TODO: map global heap data to raw data here? */
```

Comment:
Not sure if global heap data needs to be mapped to raw data for each type entry here, or if that's not appropriate?

Reply:
Presumably that would be done in H5C.c, not here. The cache client should report what type of data it is if it doesn't already.

Reply:
The remapping is done in H5F_block_write(), which we are replacing with this function, so perhaps that should be done here as well?

Reply:
No discernable performance difference.

[summit_pr.pdf](https://github.com/HDFGroup/hdf5/files/9160553/summit_pr.pdf)

Reply:
> I suppose you could set it to MEM_GHEAP in H5C.c then scan the array here to make it like selection and scalar I/O, but I'd try to come up with a consistent way to avoid a second pass over that array. We should look into what exactly these types are used for and think if we really need to change them around or just pass the real values to the file driver.

At least when the write comes from H5Cmpio.c, the type should always be H5FD_MEM_DEFAULT, so it shouldn't (at least I think) be a concern for the current PR's changes. For future things making calls to H5F_shared_vector_write though, it might be important to catch MEM_GHEAP; that's what I'm not sure of. Maybe it makes more sense for the upper layers to do the translation before making this call to avoid the extra pass over the vectors like you mention. I'm assuming the translation was simply added to H5F_block_write/read to try to keep it to one place, but that's a little more complicated of a question if you have large I/O vectors.